x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/ToCharFormatter.java

```diff
@@ -0,0 +1,371 @@
+            ToCharFormatter.of("rm")
+                .formatFn("MM", month -> String.format(Locale.ROOT, "%-4s", toRome(Integer.parseInt(month)).toLowerCase(Locale.ROOT)))
+                .text(),
+            ToCharFormatter.of("TZ").formatFn(ToCharFormatter::zoneAbbreviationOf).text(),
+            ToCharFormatter.of("tz").formatFn(t -> zoneAbbreviationOf(t).toLowerCase(Locale.ROOT)).text(),
+            ToCharFormatter.of("TZH").lowercaseAccepted(false).formatFn("ZZ", s -> s.substring(0, 3)).text(),
+            ToCharFormatter.of("TZM").lowercaseAccepted(false).formatFn("ZZ", s -> lastNCharacter(s, 2)).text(),
+            ToCharFormatter.of("OF")
+                .lowercaseAccepted(false)
+                .formatFn("ZZZZZ", ToCharFormatter::formatOffset)
+                .text(ToCharFormatter::fillOffset)).stream().map(Builder::build).collect(Collectors.toList());
+
+        Map<String, ToCharFormatter> formatterMap = new LinkedHashMap<>();
+        for (ToCharFormatter formatter : formatters) {
+            formatterMap.put(formatter.pattern, formatter);
+        }
+        // also index the lower case version of the patterns if accepted
+        for (ToCharFormatter formatter : formatters) {
+            if (formatter.lowercaseAccepted) {
+                formatterMap.putIfAbsent(formatter.pattern.toLowerCase(Locale.ROOT), formatter);
+            }
+        }
+        FORMATTER_MAP = formatterMap;
+    }
+
+    @SuppressWarnings("OptionalGetWithoutIsPresent") private static final int
+        MAX_TO_CHAR_FORMAT_STRING_LENGTH =
+        FORMATTER_MAP.keySet().stream().mapToInt(String::length).max().getAsInt();
+
+    private final String pattern;
+    private final boolean lowercaseAccepted;
+    private final Function<String, String> fillMode;
+    private final boolean hasOrdinalSuffix;
+    private final Function<TemporalAccessor, String> formatFn;
+
+    private ToCharFormatter(
+        String pattern,
+        boolean lowercaseAccepted,
+        Function<String, String> fillMode,
+        boolean hasOrdinalSuffix,
+        Function<TemporalAccessor, String> formatFn) {
+
+        this.pattern = pattern;
+        this.lowercaseAccepted = lowercaseAccepted;
+        this.fillMode = fillMode;
+        this.hasOrdinalSuffix = hasOrdinalSuffix;
+        this.formatFn = formatFn;
+    }
+
+    private static Builder of(String pattern) {
+        return new Builder(pattern);
+    }
+
+    private String format(TemporalAccessor temporalAccessor) {
+        return formatFn.apply(temporalAccessor);
+    }
+
+    private ToCharFormatter withModifier(Function<String, String> modifierFn) {
+        return new ToCharFormatter(pattern, lowercaseAccepted, fillMode, hasOrdinalSuffix, formatFn.andThen(modifierFn));
+    }
+
+    private static List<ToCharFormatter> parsePattern(String toCharPattern) {
+        LinkedList<ToCharFormatter> formatters = new LinkedList<>();
+
+        boolean fillModeModifierActive = false;
```

Comment:
Couldn't this be moved inside the while loop, reset to false on every iteration, like `foundPattern`? It currently seems reset whether a pattern is found or not.

Reply:
No, I cannot. If I move it into the `while` loop it will reset to `false` right after I have found the `FM` (the `for` exits and would reset this boolean to `false` before I find the next pattern the `FM` should act on).